### PR TITLE
support OP_MSG in authentication and runCommand

### PR DESCRIPTION
**Pull request by WebFreak001**
currently this only first adds OP_MSG support, it's not used anywhere except in the handshake and running commands. I have included CRUD struct definitions from vibe.d for a future CRUD implementation here as well.

We can wait with this PR until I added the new CRUD APIs to actually be able to deprecate the old query, etc. methods as they won't start working with this change. But I think this is good as a base.

Additionally we should probably add more integration tests to see if inserting and querying actually works as expected.

### CODE REVIEW

**Comment by adamdruppe** [comment]
yeah i run the tests just by hand with the -version=Demo thing since it needs a server setup. but yeah let's take a look at it.

File: LICENSE [hunk 1]
```diff
@@ -1,4 +1,11 @@
-Copyright (c) 2020 Symmetry Investments, with portion (c) 2012-2016 Nicolas Gurrola
+Parts of the code (identified by comments within) fall within these other licenses:
+
+Copyright (c) 2014-2022 MongoDB specification maintainers
```
Comment (adamdruppe) [inline LICENSE:3]: it is the crud.d that does this right? this license seems to prohibit symmetry from using it.....

**Comment by adamdruppe** [comment]
that said it passed all my tests so i can merge immediately and start taking it from here just that license thing is a potential problem given the purpose of this thing

**Comment by WebFreak001** [comment]
not sure what to do with the CRUD definitions - one needs to read that driver documentation, which is licensed under the here linked license. I interpreted what was written in there and based on that I written my CRUD options definition structs. But I think in other drivers they basically have the definitions too, but they are probably all licensed differently... (at least not using creative commons)

**Comment by adamdruppe** [comment]
If it is based on the documentation instead of copy/pasted from their code, their license doesn't apply - this is what mingw did with the Windows api. They read the docs and made code that implemented what was described there, and since that wasn't based on Microsoft's own code, it is considered a separate implementation.

**Comment by WebFreak001** [comment]
A problem here is that they write their documentation in a way where they tell how drivers should design their API as basically code blocks, e.g. https://github.com/mongodb/specifications/blob/master/source/crud/crud.rst#api

They state that the docs in there are for driver developers and should not be put into the user-facing docs of the drivers.

What I mostly did was copy-pasting these definitions, convert them to valid D code, removing driver internal stuff from the docs, annotate with metadata that might be relevant for the user and checked and rewrote the docs, so they are all relevant for the library user.

**Comment by adamdruppe** [comment]
Well, after looking into it, the api itself is separate from any implementation of it, so that license applies to copies of the documentation, not the code itself, which isn't compiled into the binary. The other mongo drivers can use the apache license among others https://www.mongodb.com/community/licensing so just need to make this clear. I'll do the edits.

**Pull request merged**